feat(purchase): remove unconfirmed purchase orders

jniles opened the pull request:
Adds the ability to delete unconfirmed purchase orders if they are not confirmed.  This is only active if the general "allow record deletion" ability is turned on.  It will only be useful when #5379 is implemented.

I've also renamed "purchase/list" to "purchase/registry" since it is incredible confusing as named.

Closes #5490.

Here is what it looks like in action:
![WtP4UuoSlc](https://user-images.githubusercontent.com/896472/111084530-24034900-8513-11eb-9644-954c5488cc6a.gif)

jniles committed:
feat(purchase): remove unconfirmed purchase orders

Adds the ability to delete unconfirmed purchase orders if they are not
confirmed.  This is only active if the general "allow record deletion"
ability is turned on.  It will only be useful when #5379 is implemented.

Closes #5490.

jmcameron commented on server/controllers/finance/purchases.js:
```diff
@@ -533,6 +535,31 @@ function purchaseBalance(req, res, next) {
+  try {
+    const record = await db.one('SELECT * FROM purchase WHERE uuid = ?', pid);
+
+    const WAITING_CONFIRMATION = 1;
```
Good check!

jmcameron commented on server/controllers/finance/purchases.js:
```diff
@@ -533,6 +535,31 @@ function purchaseBalance(req, res, next) {
+
+    const WAITING_CONFIRMATION = 1;
+    if (record.status_id !== WAITING_CONFIRMATION) {
+      throw new BadRequest('Cannot remove a purchase order that is confirmed');
```
Since I presume the status_id could contain more than 2 values (otherwise it would be a boolean, right?), it seems like this bad request warning should be something like:  "Can only remove purchases that have not been confirmed".

jniles replied:
Yeah, this a good suggestion.  I'll make the change.

jmcameron reviewed (approved):
Looks good.

I did verify that the cascading deleted all the purchase order items associated with this PR (via direct SQL queries).    I wonder if it is worth checking that in the integration test?

jniles committed:
test(purchases): unit test FK link

This commit adds a unit test to ensure the foreign keys work and that
values are deleted from the purchase_item table when they are removed
from the purchase table.

jmcameron commented on server/config/routes.js:
```diff
@@ -669,6 +669,7 @@ exports.configure = function configure(app) {
   app.get('/purchases/purchaseState', purchases.purchaseState);
   app.get('/purchases/:uuid', purchases.detail);
   app.put('/purchases/:uuid', purchases.update);
+  app.delete('/purchases/:uuid/', purchases.remove);
```
Do you need/want the trailing slash in the URL?

jniles replied:
Eh, it doesn't make any difference.  ExpressJS has an option to ignore them by default.

jmcameron commented on test/server-unit/purchases/purchases.spec.js:
```diff
@@ -0,0 +1,42 @@
+    const purchases = await db.exec(`SELECT * FROM purchase WHERE uuid = HUID('${puid}');`);
+    expect(purchases).to.have.length(0);
+    const items = await db.exec(`SELECT * FROM purchase_item WHERE purchase_uuid = HUID('${puid}');`);
+    expect(items).to.have.length(0);
```
Nice!  Glad to see that you verify that the PR items also get deleted.   It may not be necessary because the SQL table definitions enforce that, but I can imaging someone modifying the table definitions in the future and messing this up.  So I think this unit test is worth doing.

jniles replied:
Yeah, I agree.  I didn't really want to do it, but the peace of mind is well worth it.

jmcameron reviewed (approved):
Looks good.

I ran the tests on my PC and the server, client, and integration unit tests ran fine.   The end-to-end test failed in the same way that master does currently (deleting employee fails).
